<!-- ロボットの名前 -->
<robot name="diff_drive_robot">
  <!-- ロボットの基準とするベースリンク -->
  <link name="base_link"/>

  <!-- ボディの作成 -->
  <link name="frame">
    <visual>
      <geometry>
        <cylinder radius="0.2" length="0.15"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <cylinder radius="0.2" length="0.15"/>]
      </geometry>
    </collision>
    <!--慣性-->
    <inertial>
        <mass value="1.0"/>
        <inertia ixx="0.011875000000000002" iyy="0.011875000000000002" izz="0.020000000000000004" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>

  <!-- タイヤの作成 -->
  <link name="wheel_left">
    <visual>
      <geometry>
        <cylinder radius="0.15" length="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="blue">
        <color rgba="0.0 0.0 1.0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <cylinder radius="0.15" length="0.05"/>]
      </geometry>
    </collision>
    <!--慣性-->
    <inertial>
        <mass value="0.5"/>
        <inertia ixx="0.002916666666666667" iyy="0.002916666666666667" izz="0.005625" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>


  <!-- タイヤの作成 -->
  <link name="wheel_right">
    <visual>
      <geometry>
        <cylinder radius="0.15" length="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="cyan">
        <color rgba="0.0 1.0 1.0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <cylinder radius="0.15" length="0.05"/>
      </geometry>
    </collision>
    <!--慣性-->
    <inertial>
        <mass value="0.5"/>
        <inertia ixx="0.002916666666666667" iyy="0.002916666666666667" izz="0.005625" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>


  <!-- キャスターの作成 -->
  <link name="caster_forward">
    <visual>
      <geometry>
        <sphere radius="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="green">
        <color rgba="0.0 1.0 0.0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <sphere radius="0.05"/>
      </geometry>
    </collision>
    <!--慣性-->
    <inertial>
        <mass value="0.5"/>
        <inertia ixx="3.000e-4" iyy="3.000e-4" izz="3.000e-4" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>
  
  <!-- キャスターの作成 -->
  <link name="caster_backward">
    <visual>
      <geometry>
        <sphere radius="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="black">
        <color rgba="0.0 0.0 0.0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <sphere radius="0.05"/>
      </geometry>
    </collision>
    <!--慣性-->
    <inertial>
        <mass value="0.5"/>
        <inertia ixx="3.000e-4" iyy="3.000e-4" izz="3.000e-4" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>

  <!-- Lidar台の作成 -->
  <link name="lidar_base">
    <visual>
      <geometry>
        <box size="0.1 0.1 1.0"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="yellow">
        <color rgba="1.0 1.0 0.0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </collision>
    <!--慣性-->
    <inertial>
        <mass value="0.15"/>
        <inertia ixx="0.25" iyy="0.25" izz="0.25" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>

  <!-- Lidarの作成 -->
  <link name="laser">
    <visual>
      <geometry>
        <cylinder radius="0.05" length="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="purple">
        <color rgba="1.0 0.0 1.0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <cylinder radius="0.05" length="0.05"/>
      </geometry>
    </collision>
    <!--慣性-->
    <inertial>
        <mass value="0.15"/>
        <inertia ixx="0.125" iyy="0.125" izz="0.188" ixy="0" ixz="0" iyz="0"/>
    </inertial>
  </link>


  <link name="livox_frame"/>
  <link name="imu_link"/>
  <link name="line_link"/>
  <link name="pcl_link"/>

<!-- ################# ジョイント設定 ################## -->

	<joint name="base_joint" type="fixed">
    <origin xyz="0 0 0.15" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="frame"/>
 	</joint>

	<joint name="wheel_left_joint" type="continuous">
    <axis xyz="0 0 1"/>
    <origin xyz="0 0.225 0" rpy="1.57 0 0"/>
    <parent link="frame"/>
    <child link="wheel_left"/>
 	</joint>

	<joint name="wheel_right_joint" type="continuous">
    <axis xyz="0 0 1"/>
    <origin xyz="0 -0.225 0" rpy="1.57 0 0"/>
    <parent link="frame"/>
    <child link="wheel_right"/>
	</joint>

	<joint name="caster_forward_joint" type="fixed">
    <origin xyz="-0.15 0 -0.1" rpy="0 0 0"/>
    <parent link="frame"/>
    <child link="caster_forward"/>
	</joint>

	<joint name="caster_backward_joint" type="fixed">
    <origin xyz="0.15 0 -0.1" rpy="0 0 0"/>
    <parent link="frame"/>
    <child link="caster_backward"/>
	</joint>

	<joint name="lidar_base_joint" type="fixed">
    <origin xyz="0.0 0.0 0.5" rpy="0 0 0"/>
    <parent link="frame"/>
    <child link="lidar_base"/>
	</joint>

	<joint name="lidar_joint" type="fixed">
    <origin xyz="0.0 0.0 1.0" rpy="0 0 0"/>
    <parent link="frame"/>
    <child link="laser"/>
	</joint>



  <joint name="livox_joint" type="fixed">
    <parent link="laser"/>
    <child link="livox_frame"/>
    <origin xyz="0.0 0.0 0.0" rpy="0.0 0.6108 0.0"/>
  </joint>

  <joint name="imu_joint" type="fixed">
    <parent link="laser"/>
    <child link="imu_link"/>
    <origin xyz="0.0 0.1 0.0" rpy="0.0 3.14 0.0"/>
  </joint>

  <joint name="line_joint" type="fixed">
    <parent link="base_link"/>
    <child link="line_link"/>
    <origin xyz="0.0 0.0 0.1" rpy="0.0  0.0 -1.57"/>
  </joint>

  <joint name="pcl_joint" type="fixed">
    <parent link="base_link"/>
    <child link="pcl_link"/>
    <origin xyz="0.0 0.0 0.0" rpy="0.0  0.0 -3.14"/>
  </joint>
</robot>
